lint: Run the CI lint stage on mac

Empact opened the pull request:
This helps ensure ongoing compatibility with macOS-distributed version of GNU bash.

Empact commented:
Inspired by and built on #13720

fanquake commented on .travis.yml:
```diff
@@ -105,3 +110,34 @@ jobs:
           fi
+    - stage: lint BSD
+      os: osx
+      # Use macOS consistent with the minimum supported version: 10.11
```
If you mean our next supported minimum macOS version, shouldn't this be 10.10?

Empact replied:
Thanks - was relying on faulty memory. Also it happens to be the earliest version supported by Travis. Updated. #13617

practicalswift commented:
Concept ACK.

Nit: macOS ships with GNU bash, albeit an ancient version. Please change "helps ensure ongoing compatibility with BSD bash as well as GNU bash" to something along the lines of "helps ensure ongoing compatibility with the macOS BSD user land and the ancient version of GNU bash shipped with macOS".

sipa commented:
```
/usr/local/Homebrew/Library/Homebrew/brew.rb:12:in `<main>': Homebrew must be run under Ruby 2.3! You're running 2.0.0. (RuntimeError)
```

Empact commented:
wip: I’m rolling back to 10.10

MarcoFalke commented:
Might as well do a full depends and bitcoind build when pulling in macos anyway?

scravy commented on .travis.yml:
```diff
@@ -1,6 +1,8 @@
-os: linux
+os:
+  - linux
+  - osx
```
AFAIK it is not required to enumerate `osx` here.

scravy commented on .travis.yml:
```diff
@@ -11,6 +13,7 @@ cache:
 stages:
   - lint
   - test
+  - lint macOS
```
you can use the same stage `lint` - it will have this stage with two jobs, one in linux, one in macOS

scravy commented on .travis.yml:
```diff
@@ -81,6 +84,8 @@ after_script:
 
 jobs:
+  exclude:
+    - os: osx
```
this would not be required either if not mentioning it in the `os` section globally

scravy commented on .travis.yml:
```diff
@@ -105,3 +110,41 @@ jobs:
               travis_wait 50 contrib/verify-commits/verify-commits.py;
           fi
+    - stage: lint macOS
+      os: osx
```
this `osx` suffices

scravy commented on .travis.yml:
```diff
@@ -105,3 +110,41 @@ jobs:
               while read LINE; do travis_retry gpg --keyserver hkp://subset.pool.sks-keyservers.net --recv-keys $LINE; done < contrib/verify-commits/trusted-keys &&
               travis_wait 50 contrib/verify-commits/verify-commits.py;
           fi
+    - stage: lint macOS
```
this could be `lint` as mentioned above

scravy reviewed (changes requested):
I very much like these changes - love the `--exclude` change, the `-e` made me go to the documentation the first time I came across it and I wished it was a more self-explanatory cli switch :-)

scravy commented:
ACK d5d76655105d6d338dbc367ae0b54b7433ffeb38

practicalswift commented:
utACK d5d7665

Closing my PR #13720 in favour of this one.

MarcoFalke commented:
There are some non-travis related changes here, which should be submitted independently, imo.

Empact commented:
On second thought I instead set up CI runs against this PR with each of the prior commits excluded, to demonstrate the associated failure:
sans 4874450 / empty subexpressions: https://travis-ci.org/Empact/bitcoin/builds/421218998
sans 28d520b / SC1087: https://travis-ci.org/Empact/bitcoin/builds/421219652
sans a442c33 / SC1117: https://travis-ci.org/Empact/bitcoin/jobs/421201432
sans 21ff32e / SC2206: https://travis-ci.org/Empact/bitcoin/builds/421219244
sans 2917ecd / SC2207: https://travis-ci.org/Empact/bitcoin/builds/421220532
sans f65efb2 / SC2230: https://travis-ci.org/Empact/bitcoin/builds/421219961

Empact commented:
Rebased for #14115. This now only adds the mac-based lint run to CI.

ken2812221 commented:
utACK 453daa0.
Maybe put it in test stage. It takes 3 minutes to boot and 4 minutes to run the script.

practicalswift commented:
utACK 453daa0e24ae441175f413dd03460417eb4f4ee5

Will be great to have this in Travis!

nit: Remove Linux in PR title? :-)

Empact commented:
Moved it into a 3rd/final "Compat" stage. Sound good?

scravy commented on .travis.yml:
```diff
@@ -135,3 +135,24 @@ jobs:
+      # https://docs.travis-ci.com/user/reference/osx/#OS-X-Version
+      osx_image: xcode7.3
+      # TODO: if you're updating osx_image, try using rvm: to supply the version of ruby
+      # required by homebrew. Could not get this working with xcode7.3 but believe it should
```
It says `osx_image: xcode7.3` and at the same time `could not get this working with xcode7.3`

MarcoFalke commented:
Should be marked "up for grabs"?

fanquake commented:
@Empact Looks like you've rebased since Marco's comment? Travis had errored out, so I restarted, but it now seems to be stalled while installing dependencies with `brew`?

Empact commented:
@fanquake FYI I'm picking this back up.

Empact committed:
lint: Fix shellcheck SC2155

Declare and assign separately to avoid masking return values.

https://github.com/koalaman/shellcheck/wiki/SC2155

Empact commented:
Currently based on #15257 in order to give myself a clean build locally. Happy to base otherwise.

Empact commented:
Outstanding work: need to install git with USE_LIBPCRE

> fatal: cannot use Perl-compatible regexes when not compiled with USE_LIBPCRE

Empact committed:
lint: Install grep and git via brew on mac for --perl-regexp

Particularly `--with-pcre2` is needed to run `git grep --perl-regexp` in
`test/link/check-doc.py`

Empact committed:
lint: Run the linters against Mac OS on Travis

This helps ensure ongoing compatibility with macOS-distributed version
of GNU bash.

Sjors commented:
ACK eefbc9d. Looking forward to seeing more stuff tested on the macOS native build.

fanquake commented on .travis.yml:
```diff
@@ -99,6 +100,30 @@ jobs:
+        - 2.3.7
+      # clear default env
+      env:
+      sudo: false
```
I'm pretty sure you can drop sudo here.

fanquake commented on .travis/lint_04_install.sh:
```diff
@@ -6,10 +6,23 @@
+if [ "$TRAVIS_OS_NAME" == "osx" ]; then
+  # update first to install required ruby dependency
+  travis_retry brew update
+  travis_retry brew reinstall git -- --with-pcre2 # for  --perl-regexp
```
I don't think `--with-pcre2` is a `git` install option anymore? Or is this being passed because we're using an old version of `brew` or something?

Empact replied:
[This build](https://travis-ci.org/bitcoin/bitcoin/jobs/595626506) demonstrates `--with-pcre2` is needed to run `git grep --perl-regexp` in `test/link/check-doc.py`.

fanquake reviewed (commented):
I feel like I've Concept ACK'd this already, but re-Concept ACK anyways. Left a couple comments inline. It might be handy to base this off #16582, to make it easier to test.

MarcoFalke commented on ci/lint/04_install.sh:
```diff
@@ -6,10 +6,23 @@
 
+if [ "$TRAVIS_OS_NAME" == "osx" ]; then
+  # update first to install required ruby dependency
+  travis_retry brew update
```
style-nit: Could use `./ci/retry` to not depend on travis?

jonasschnelli commented on ci/lint/04_install.sh:
```diff
@@ -6,9 +6,22 @@
+  export PATH
+else
+  SHELLCHECK_VERSION=v0.6.0
+  curl -s "https://storage.googleapis.com/shellcheck/shellcheck-${SHELLCHECK_VERSION}.linux.x86_64.tar.xz" | tar --xz -xf - --directory /tmp/
```
Probably a good idea to add a `travis_retry` here as well

Empact replied:
Added `travis_retry` and switched `-s` to `--silent` for clarity.

@MarcoFalke will plan on a follow-up converting the lint scripts

MarcoFalke replied:
Ah, looks like setup_env is already running (because the `before_install` step is inherited.

https://travis-ci.org/bitcoin/bitcoin/jobs/595662274#L125

MarcoFalke commented on ci/lint/04_install.sh:
```diff
@@ -6,9 +6,22 @@
-SHELLCHECK_VERSION=v0.6.0
-curl -s "https://storage.googleapis.com/shellcheck/shellcheck-${SHELLCHECK_VERSION}.linux.x86_64.tar.xz" | tar --xz -xf - --directory /tmp/
-export PATH="/tmp/shellcheck-${SHELLCHECK_VERSION}:${PATH}"
+travis_retry pip3 install vulture==0.29
```
This has been removed, no?

fanquake replied:
Indeed. Removed in #16961.

Empact replied:
Indeed, removed in #16961, and here now too. :P

MarcoFalke commented:
I am going to merge this under the "experimental umbrella", as it might be needed for  #16597

MarcoFalke commented:
It takes about 10 minutes to boot the macOS on travis. If that is too long, we can revert this.

ACK cd82f75a431967030d25bcfe7c4aaefe5b345a99

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK cd82f75a431967030d25bcfe7c4aaefe5b345a99
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUij+wv8DunewWxtXKFPL59lT0GJQYyfc4ueELUQTYrEf8pyG9Ryd6FPs4jd67u0
eb6D4q7M/8vZFt9TfO5MfhG9CVNEhV7sY6N1b8iXKCl7giyI8eezo2/QLYInrbFs
/zpFbu0jlgcQUC8UjsfZjdckmyIKgFQtsM5j0aXDgkTxdgOOTztNHPw4J8HVypvX
kzrZXg7gN63RgEhS08toErpAKtIssahjn8swpks7iZpsckNUmwuA4qV5251K+ZcV
phd8lViW5VkH9Gk75gDo1UThFLs4Fpg2xN9JVG4sKi1vSFpdhZV+fvPfkwWl152p
hXmfwmn+UGlKnzLhgZNMYirBSkS5EEtBQSM1nlpR48+G9ABCCeRnehFchfC/DWSu
585fGTGyHQ1JJn64JnxeAw2ECURMn0t7Gmbr/A4pXUNGnunWhMQvOq3LD67iXrHV
KzYV3pGcXXn2bSJUEOoWPfFJi5Vi2likTTbFfRd6rGlxX1hqUZFfuTgWoFdDGTGx
1HiNi67t
=EoXU
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `83acb4bc3587c0498eb8a1fc412634b436d9456759eec0634bddf78144e5b4e4  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010883acb4bc3587c0498eb8a1fc412634b436d9456759eec0634bddf78144e5b4e4f010f586b95ffd253bc736335a57ce0d2dd308fff010179b915b096e5bdcf1b4944708594d2108f1045d9e27cbf008575e308b039fdf810083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0103c3eeae68fe1055360e1fa8f2545bace08f1045d9e27cbf008555c061ffd3150b60083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff0105141a98c91ac4af768a275e83cd05fee08f1045d9e27cbf008b697608b1dd52ca10083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f0105ae333ba6a2bc132a6a1bd6632264bf108f1045d9e27cbf00866a3655773d22a120083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

fanquake commented:
> It takes about 10 minutes to boot the macOS on travis.

@MarcoFalke Any idea why it takes that long?

MarcoFalke commented:
> @MarcoFalke Any idea why it takes that long?

because travis 🤷

MarcoFalke commented:
The backlog for macOS can be seen here: https://www.traviscistatus.com/

note to myself: It failed here: https://travis-ci.org/bitcoin/bitcoin/jobs/595893060#L2470

ryanofsky commented:
I think it's bad that this change seems to now require scripted diffs to run in the macos unix environment. It makes me less inclined to write scripted diffs if I have to somehow get them them to work in macos when I don't use the operating system, and don't have the hardware, and all the unix tools are strange and old.

Also, does this scripted-diff checking script itself not function on macos...?

https://travis-ci.org/bitcoin/bitcoin/jobs/598935472

```sh
$ set -o errexit; source ./ci/lint/06_script.sh
sed: 1: "/^-BEGIN VERIFY SCRIPT- ...": unexpected EOF (pending }'s)
Error: missing script for: 3c5edd40ae63daee4de6c58ea4123c66be83119c
Failed
sed: 1: "/^-BEGIN VERIFY SCRIPT- ...": unexpected EOF (pending }'s)
Error: missing script for: 7e271ea765b731c59edbc62790f568af284c7800
Failed
```

Sjors commented:
There are some tools / homebrew packages to make macOS scripts behave more like Linux. E.g. https://formulae.brew.sh/formula/gnu-sed

MarcoFalke commented:
Ok, then lets remove this and macOS users should use the gnu untils that are provided by brew
